Training24/NumConversion.c

```diff
@@ -0,0 +1,100 @@
+// ------------------------------------------------------------------------------------------------
+// Training ~ A training program for new joiners at Metamation, Batch - July 2024.
+// Copyright (c) Metamation India.
+// ------------------------------------------------------------------
+// Numconversion.c
+// Program to get user input(only integer) and convert them to Hexadecimal and Binary.
+// ------------------------------------------------------------------------------------------------
+
+#include <stdio.h>
+#include <stdbool.h>
+#include <string.h>
+#include <ctype.h>
+#include <limits.h>
+
+/// <summary>Function to  convert an integer to binary.</summary>
+void ConvertToBinary (int number, int binaryArray[], int bitWidth) {
+   for (int i = bitWidth - 1; i >= 0; i--) binaryArray[bitWidth - 1 - i] = (number >> i) & 1;
+   return;
+}
+
+/// <summary>Function to print the binary representation from the binary array.</summary>
+void PrintBinary (const int binaryArray[], int bitWidth) {
+   for (int i = 0; i < bitWidth; i++) printf ("%d", binaryArray[i]);
+   printf ("\n");
+}
+
+/// <summary>Function to  check if input is a valid integer.</summary>
+bool IsValidInteger (const char* str) {
+   if (*str == '-' || *str == '+') str++;   // Skip sign if present
+   if (!*str) return false;   // If string is empty after sign
+   while (*str) {   // Check each character to ensure it's a digit
+      if (!isdigit (*str)) return false; 
+      str++;
+   }
+   return true;
+}
+
+/// <summary>Function to get a valid integer from the user.</summary>
+bool GetValidInteger (int* number) {
+   char input[100];
+   if (fgets (input, sizeof (input), stdin) != NULL) {
+      input[strcspn (input, "\n")] = '\0';     // Remove newline character if present
+      if (IsValidInteger (input) && sscanf_s (input, "%d", number) == 1) return true;
+   }
+   return false;
+}
+
+/// <summary>Function to convert an integer into hexadecimal representation.</summary>
+void ConvertToHex (int number, char hex[], int bitWidth) {
+   int value;
+   int hexDigits = bitWidth / 4;   // Calculate number of hex digits based on bit width
+   // Handle negative numbers by computing two's complement
+   if (number < 0) value = (1 << bitWidth) + number;
+   else value = number;
+   int index = 0;
+   do {
+      int remainder = value % 16;
+      if (remainder < 10) hex[index++] = '0' + remainder;
+      else hex[index++] = 'A' + (remainder - 10);
+      value /= 16;
+   } while (value > 0);
+   while (index < hexDigits) hex[index++] = '0';   // Pad with leading zeros for the required bit width

```

Comment:
You could also fill the array in reverse order, so you don't have to traverse it twice. int index = 0, hexDigits = bitWidth / 4, remainder = 0, 
    value = number < 0 ? (1 << bitWidth) + number : number;
```suggestion
hex[hexDigits] = '\0';
do {
   remainder = value % 16;
   hex[--hexDigits] = remainder < 10 ? '0' + remainder : 'A' + (remainder - 10);
   value /= 16;
} while (value > 0);
```